Simulation userguide (#172)
emdgroup/baybe: docs/userguide/simulation.md

# Simulation

BayBE offers multiple functionalities to "simulate" experimental campaigns with a given lookup mechanism. This user guide briefly introduces how to use the methods available in our [simulation submodule](baybe.simulation).

For a wide variety of applications of this functionality, we refer to the corresponding [examples](../../examples/Backtesting/Backtesting).

## Terminology: What do we mean by "Simulation"?

The term "simulation" can have two slightly different interpretations, depending on the applied context.

1. It can refer to "backtesting" a particular experimental campaign on a fixed finite dataset.
In this context, "simulation" means investigating what experimental trajectory we would have observed if we had applied the recommender in a certain defined context and restricted the possible parameter configurations to those contained in the dataset.

> 2. It can refer to the simulation of an *actual* DOE loop, i.e., recommending experiments and retrieving the corresponding measurements, where the loop closure is realized in the form of a callable (black-box) function that can be queried during the optimization to provide target values.

Reviewer: add
`Such callable could for instance be a simple analytical function or numerical solver of a set of differential equations that describe a physical system.` (or similar)

> In BayBE, the simulation submodule allows a wide range of use cases and can even be used for "oracle predictions".

Reviewer: feels like this last paragraph should already be written below the next subheading `The lookup functionality`

Author: Which paragraph exactly do you mean here? The one starting with "In BayBE"?

Reviewer: yes, the last paragraph before the mentioned heading

Author: Changed.

This is enabled by the proper use of the `lookup` functionality, which allows to either use fixed data sets, analytical functions, and general callbacks for retrieving target function values. 

## The lookup functionality

All functions require a `lookup` which is used to close the loop and return target values for points in the search space.
It can be provided in the form of a dataframe or a `Callable`.

```{note}

> Technically, the `lookup` can also be `None`. This results in the simulation producing fake results which is not useful for real applications and thus not discussed further.

Reviewer: `not useful` is maybe a bit to negatively worded here, it is for instance useful for simply testing code before you care about implementing the lookup. maybe just remove that subsentence so it says plain `fake results, which is not discussed further.` or beautify it with HIC

also prob replace `fake` by `random`

```

### Using a dataframe

> When choosing a dataframe, the dataframe needs to contain experimental settings and their target results.

Reviewer: ```suggestion
When choosing a dataframe, it needs to contain parameter combinations and their target results.
```

> An important aspect to keep in mind when providing a dataframe is that one either needs to ensure that **every** possible combination of parameters is part of the dataframe or that it can happen that a recommendation is made for which the target value cannot be found in the dataframe.

Author: Rewrote, please resolve if gucci.

~~~python
results = simulate_scenarios(
    # Necessary
    campaign=campaign,
    # Technically optional but should always be set
    lookup=lookup,
    # Optional
    batch_size=batch_size,
    n_doe_iterations=n_doe_iterations,
    initial_data=initial_data,
    random_seed=random_seed,
    impute_mode=impute_mode,
    noise_percent=noise_percent,
)
~~~

This function returns a dataframe that contains the results. For details on the columns of this dataframe as well as the dataframes returned by the other functions discussed here, we refer to the documentation of the submodule [here](baybe.simulation).

## Simulating multiple scenarios

The function [`simulate_scenarios`](baybe.simulation.simulate_scenarios) allows to specify multiple simulation settings at once.
Instead of a single campaign, this function expects a dictionary of campaigns, mapping scenario identifiers to `Campaign` objects.

> In addition to the keyword arguments available for `simulate_experiment`, this function has the additional keyword argument `n_mc_iterations`. This keyword argument can be used to perform multiple Monte Carlo runs with  single call.

Author: Fixed the formulation, please resolve if gucci.
